packages/commonwealth/client/scripts/views/pages/contracts/add_contract_form.tsx

```diff
@@ -0,0 +1,180 @@
+/* @jsx m */
+
+import m from 'mithril';
+import $ from 'jquery';
+
+import 'pages/create_community.scss';
+
+import app from 'state';
+import {
+  ChainBase,
+  ChainNetwork,
+  ChainType,
+  ContractType,
+} from 'common-common/src/types';
+import { isAddress } from 'web3-utils';
+
+import { notifyError } from 'controllers/app/notifications';
+import { IdRow, InputRow, SelectRow } from 'views/components/metadata_rows';
+
+import { CWButton } from 'views/components/component_kit/cw_button';
+import { CWValidationText } from 'views/components/component_kit/cw_validation_text';
+
+import { linkExistingAddressToChainOrCommunity } from 'controllers/app/login';
+
+import {
+  initChainForm,
+  defaultChainRows,
+  ethChainRows,
+} from '../create_community/chain_input_rows';
+
+import {
+  ChainFormFields,
+  ChainFormState,
+  EthChainAttrs,
+  EthFormFields,
+} from '../create_community/types';
+
+type ContractFormFields = {
+  chain_node_id: number;
+  abi: JSON;
```

Comment:
What is "JSON" type? Don't web3 libs have an ABI type?

Reply:
They do not, the best we could do is have it as an array of AbiItems, so I'll do that!

Reply:
or I can have them as an array of Records?